src/main/java/com/commercetools/sync/producttypes/helpers/ProductTypeBatchProcessor.java

```diff
@@ -0,0 +1,151 @@
+package com.commercetools.sync.producttypes.helpers;
+
+import com.commercetools.sync.commons.exceptions.InvalidProductTypeDraftException;
+import com.commercetools.sync.commons.exceptions.InvalidReferenceException;
+import com.commercetools.sync.producttypes.ProductTypeSync;
+import io.sphere.sdk.products.attributes.AttributeDefinitionDraft;
+import io.sphere.sdk.products.attributes.AttributeType;
+import io.sphere.sdk.products.attributes.NestedAttributeType;
+import io.sphere.sdk.products.attributes.SetAttributeType;
+import io.sphere.sdk.producttypes.ProductTypeDraft;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.commercetools.sync.commons.helpers.BaseReferenceResolver.BLANK_ID_VALUE_ON_RESOURCE_IDENTIFIER;
+import static java.lang.String.format;
+import static java.util.Collections.emptySet;
+import static org.apache.commons.lang3.StringUtils.isBlank;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+
+public class ProductTypeBatchProcessor {
+    static final String PRODUCT_TYPE_DRAFT_KEY_NOT_SET = "ProductTypeDraft with name: %s doesn't have a key. "
+        + "Please make sure all productType drafts have keys.";
+    static final String PRODUCT_TYPE_DRAFT_IS_NULL = "ProductTypeDraft is null.";
+    static final String PRODUCT_TYPE_HAS_INVALID_REFERENCES = "ProductTypeDraft with key: '%s' has invalid productType "
+        + "references on the following AttributeDefinitionDrafts: %s";
+
+    private final List<ProductTypeDraft> productTypeDrafts;
+    private final ProductTypeSync productTypeSync;
+    private final Set<ProductTypeDraft> validDrafts = new HashSet<>();
+    private final Set<String> keysToCache = new HashSet<>();
+
+    public ProductTypeBatchProcessor(@Nonnull final List<ProductTypeDraft> productTypeDrafts,
+                                     @Nonnull final ProductTypeSync productTypeSync) {
+        this.productTypeDrafts = productTypeDrafts;
+        this.productTypeSync = productTypeSync;
+    }
+
+    /**
+     * This method validates the batch of drafts, and only for valid drafts it adds the valid draft
+     * to a {@code validDrafts} set, and adds the keys of those drafts and their referenced productTypes to a
```

Comment:
I see you add the referenced keys, but where do you add referenced productTypes?

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/382/files/de7ef701028b60d1670ee0df009c86d005a52028#diff-e06e39a782136b5dee4002fbd4494cb1R65